File: Programmers/[81302] 거리두기 확인하기.java
import java.util.*;

class Solution {
    static final int[] dy = {-1, 0, 1, 0};
    static final int[] dx = {0, 1, 0, -1};

    boolean isCompliedPerson(String[] place, Position applicant) {
        Queue<Position> queue = new LinkedList<>();
        int[][] visited = new int[5][5];
        for (int i = 0; i < 5; i++) {
            Arrays.fill(visited[i], -1);
        }

        visited[applicant.y][applicant.x] = 0;
        queue.offer(applicant);

        while (!queue.isEmpty()) {
            Position popped = queue.poll();
            int y = popped.y;
            int x = popped.x;

            if (visited[y][x] > 2) {
                return true;
            }

            for (int i = 0; i < 4; i++) {
                int ny = y + dy[i];
                int nx = x + dx[i];
                if (0 <= ny && ny < 5 && 0 <= nx && nx < 5) {
                    if (visited[ny][nx] == -1) {
                        char nextPosition = place[ny].charAt(nx);
                         visited[ny][nx] = visited[y][x] + 1;
                        if (nextPosition == 'P') {
                            if (visited[ny][nx] <= 2) {
                                return false;
                            }
                        }
                        if (nextPosition != 'X') {
                            queue.offer(new Position(ny, nx));
                        }
                    }
                }
            }
        }
        return true;
    }

    boolean isCompliedPlace(String[] place) {
        for (int i = 0; i < 5; i++) {
            for (int j = 0; j < 5; j++) {
                if (place[i].charAt(j) == 'P') {
                    if (!isCompliedPerson(place, new Position(i, j))) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    public int[] solution(String[][] places) {
        List<Integer> answer = new ArrayList<>();

        for (int i = 0; i < 5; i++) {
            answer.add(isCompliedPlace(places[i]) ? 1 : 0);
        }

        return answer.stream()
                .mapToInt(x -> x)
                .toArray();
    }

    static class Position {
        int y, x;

        public Position(int y, int x) {
            this.y = y;
            this.x = x;
        }
    }
}

File: Programmers/[81302] 거리두기 확인하기.py
from collections import deque

dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]


def bfs(place, y, x):
    isVisited = [[-1] * 5 for _ in range(5)]
    isVisited[y][x] = 0
    queue = deque([[y, x]])
    while queue:
        [y, x] = queue.popleft()
        if isVisited[y][x] != 0 and place[y][x] == "P":
            return False  # 거리두기 지키지 않는 경우
        if isVisited[y][x] == 2:
            continue
        for i in range(4):
            ny = y + dy[i]
            nx = x + dx[i]
            if 0 <= ny < 5 and 0 <= nx < 5:
                if isVisited[ny][nx] == -1 and place[ny][nx] != "X":
                    isVisited[ny][nx] = isVisited[y][x] + 1
                    queue.append([ny, nx])
    return True


def isFollowingRule(place):
    for i in range(5):
        for j in range(5):
            if place[i][j] == "P":
                if not bfs(place, i, j):
                    return False
    return True


def solution(places):
    answer = []
    for place in places:
        if isFollowingRule(place):
            answer.append(1)
        else:
            answer.append(0)
    return answer
